modules/snooker/src/main/ts/ephox/snooker/resize/ColumnSizes.ts

```diff
@@ -27,36 +31,52 @@ const getRawH = function (cell: SugarElement) {
 const justCols = (warehouse: Warehouse): Optional<SugarElement<HTMLTableColElement>>[] =>
   Arr.map(Warehouse.justColumns(warehouse), (column) => Optional.from(column.element));
 
-const getWidthFrom = function <T> (warehouse: Warehouse, getWidth: (cell: SugarElement, tableSize: TableSize) => T, fallback: (deduced: Optional<number>) => T, tableSize: TableSize) {
-  const columns: Optional<SugarElement>[] = Warehouse.hasColumns(warehouse) ? justCols(warehouse) : Blocks.columns(warehouse);
+// Col elements don't have valid computed widths/positions, so treat them as invalid if they don't have a raw width
+const isValidColumn = (cell: SugarElement<HTMLTableCellElement | HTMLTableColElement>) =>
+  !isCol(cell) || Css.getRaw(cell, 'width').isSome();
 
-  const backups = Arr.map(columns, function (cellOption) {
-    return cellOption.map(width.edge);
-  });
+const getDimension = <T>(cellOpt: Optional<SugarElement>, index: number, backups: Optional<number>[], filter: (cell: SugarElement) => boolean, getter: (cell: SugarElement) => T, fallback: (deduced: Optional<number>) => T): T =>
+  cellOpt.filter(filter).fold(
+    // Can't just read the width of a cell, so calculate.
+    () => fallback(Util.deduce(backups, index)),
+    (cell) => getter(cell)
+  );
+
+const getWidthFrom = function <T> (warehouse: Warehouse, table: SugarElement<HTMLTableElement>, getWidth: (cell: SugarElement, tableSize: TableSize) => T,
+                                   fallback: (deduced: Optional<number>) => T, tableSize: TableSize) {
+  const columnCells = Blocks.columns(warehouse);
+  const columns: Optional<SugarElement>[] = Warehouse.hasColumns(warehouse) ? justCols(warehouse) : columnCells;
+
+  const backups = [ Optional.some(width.edge(table)) ].concat(Arr.map(width.positions(columnCells, table), (pos) =>
+    pos.map((p) => p.x)
+  ));
+
+  // Only use the width of cells that have no column span (or colspan 1)
+  const colFilter = Fun.not(CellUtils.hasColspan);
 
   return Arr.map(columns, function (cellOption, c) {
-    // Only use the width of cells that have no column span (or colspan 1)
-    const columnCell = cellOption.filter(Fun.not(CellUtils.hasColspan));
-    return columnCell.fold(function () {
-      // Can't just read the width of a cell, so calculate.
-      const deduced = Util.deduce(backups, c);
-      return fallback(deduced);
-    }, function (cell) {
-      return getWidth(cell, tableSize);
-    });
+    return getDimension(cellOption, c, backups, colFilter, (column) => {
+      if (isValidColumn(column)) {
+        return getWidth(column, tableSize);
+      } else {
+        // Invalid column so fallback to trying to get the computed width from the cell
+        const cell = Optionals.bindFrom(columnCells[c], Fun.identity);
```

Comment:
Should this just be `Optional.from`?

Reply:
No, that'd make the type `Optional<Optional<SugarElement>>` when we want it to be `Optional<SugarElement>`